test/BRSRC13/test_register.c

```diff
@@ -1,68 +1,503 @@
 #include "tests.h"
 
 #include "CORE/FW/register.h"
+#include <string.h>
 
-void test_register_BrRegistryFind() {
+
+typedef struct {
+    br_uint_32 _reserved;
+    char *identifier;
+} test_element;
+
+static void test_register_BrRegistryNew() {
     br_registry reg;
-    br_model m;
-    br_model m2;
-    br_registry_entry* e;
-    reg.find_failed_hook = NULL;
-    m.identifier = "test-identifier1";
-    m2.identifier = "test-identifier2";
+    br_registry *result;
+
+    memset(&reg, 0, sizeof(reg));
+
+    result = BrRegistryNew(&reg);
+    TEST_ASSERT_EQUAL_PTR(&reg, result);
+    TEST_ASSERT_EQUAL_INT(0, reg.count);
+}
+
+static void test_register_BrRegistryAdd_BrRegistryRemove() {
+    br_registry reg;
+    void *result;
+
+    memset(&reg, 0, sizeof(reg));
+
+    test_element item1 = { .identifier = "item1", };
+    test_element item2 = { .identifier = "item2", };
+    test_element item3 = { .identifier = "item3", };
 
     BrRegistryNew(&reg);
-    BrRegistryAdd(&reg, &m);
-    BrRegistryAdd(&reg, &m2);
+
+    result = BrRegistryAdd(&reg, &item1);
+    TEST_ASSERT_EQUAL_PTR(&item1, result);
+    TEST_ASSERT_EQUAL_INT(1, reg.count);
+    TEST_ASSERT_EQUAL_PTR(&item1, ((br_registry_entry*)reg.list.head)->item);
+
+    result = BrRegistryAdd(&reg, &item2);
+    TEST_ASSERT_EQUAL_PTR(&item2, result);
+    TEST_ASSERT_EQUAL_INT(2, reg.count);
+    TEST_ASSERT_EQUAL_PTR(&item2, ((br_registry_entry*)reg.list.head)->item);
+
+    result = BrRegistryAdd(&reg, &item3);
+    TEST_ASSERT_EQUAL_PTR(&item3, result);
+    TEST_ASSERT_EQUAL_INT(3, reg.count);
+    TEST_ASSERT_EQUAL_PTR(&item3, ((br_registry_entry*)reg.list.head)->item);
+
+    result = BrRegistryRemove(&reg, &item1);
+    TEST_ASSERT_EQUAL_PTR(&item1, result);
+    TEST_ASSERT_EQUAL_INT(2, reg.count);
+    TEST_ASSERT_EQUAL_PTR(&item3, ((br_registry_entry*)reg.list.head)->item);
+
+    result = BrRegistryRemove(&reg, &item1);
+    TEST_ASSERT_NULL(result);
     TEST_ASSERT_EQUAL_INT(2, reg.count);
-    e = (br_registry_entry*)reg.list.head;
-    TEST_ASSERT_EQUAL_PTR(&m2, e->item);
-    TEST_ASSERT_EQUAL_STRING("test-identifier2", e->item[1]);
+    TEST_ASSERT_EQUAL_PTR(&item3, ((br_registry_entry*)reg.list.head)->item);
+
+    result = BrRegistryRemove(&reg, &item3);
+    TEST_ASSERT_EQUAL_PTR(&item3, result);
+    TEST_ASSERT_EQUAL_INT(1, reg.count);
+    TEST_ASSERT_EQUAL_PTR(&item2, ((br_registry_entry*)reg.list.head)->item);
+
+    result = BrRegistryRemove(&reg, &item2);
+    TEST_ASSERT_EQUAL_PTR(&item2, result);
+    TEST_ASSERT_EQUAL_INT(0, reg.count);
 
-    TEST_ASSERT_NOT_NULL(e->node.next);
-    e = (br_registry_entry*)e->node.next;
-    TEST_ASSERT_EQUAL_PTR(&m, e->item);
+    // Cannot execute BrRegistryRemove on an empty br_registry
```

Comment:
Comment was oudated.
This was another bug with registry: you could not remove an element when it was already empty.
```suggestion
```